File: sdks/php/src/Types/Flow.php
<?php

declare(strict_types=1);

namespace Linktor\Types;

class Flow
{
    public string $id = '';
    public string $name = '';
    public ?string $description = null;
    /** @var FlowNode[] */
    public array $nodes = [];
    /** @var FlowEdge[] */
    public array $edges = [];
    public ?FlowSettings $settings = null;
    public string $status = '';
    public ?string $publishedVersion = null;
    public array $metadata = [];
    public ?\DateTimeImmutable $createdAt = null;
    public ?\DateTimeImmutable $updatedAt = null;
    public ?\DateTimeImmutable $publishedAt = null;

    public static function fromArray(array $data): self
    {
        $instance = new self();
        $instance->id = $data['id'] ?? '';
        $instance->name = $data['name'] ?? '';
        $instance->description = $data['description'] ?? null;
        $instance->nodes = array_map(fn($n) => FlowNode::fromArray($n), $data['nodes'] ?? []);
        $instance->edges = array_map(fn($e) => FlowEdge::fromArray($e), $data['edges'] ?? []);
        $instance->settings = isset($data['settings']) ? FlowSettings::fromArray($data['settings']) : null;
        $instance->status = $data['status'] ?? '';
        $instance->publishedVersion = $data['publishedVersion'] ?? null;
        $instance->metadata = $data['metadata'] ?? [];
        $instance->createdAt = isset($data['createdAt']) ? new \DateTimeImmutable($data['createdAt']) : null;
        $instance->updatedAt = isset($data['updatedAt']) ? new \DateTimeImmutable($data['updatedAt']) : null;
        $instance->publishedAt = isset($data['publishedAt']) ? new \DateTimeImmutable($data['publishedAt']) : null;
        return $instance;
    }
}

class FlowNode
{
    public string $id = '';
    public string $type = '';
    public ?FlowNodePosition $position = null;
    public array $data = [];

    public static function fromArray(array $data): self
    {
        $instance = new self();
        $instance->id = $data['id'] ?? '';
        $instance->type = $data['type'] ?? '';
        $instance->position = isset($data['position']) ? FlowNodePosition::fromArray($data['position']) : null;
        $instance->data = $data['data'] ?? [];
        return $instance;
    }

    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'type' => $this->type,
            'position' => $this->position?->toArray(),
            'data' => $this->data,
        ];
    }
}

class FlowNodePosition
{
    public float $x = 0;
    public float $y = 0;

    public static function fromArray(array $data): self
    {
        $instance = new self();
        $instance->x = $data['x'] ?? 0;
        $instance->y = $data['y'] ?? 0;
        return $instance;
    }

    public function toArray(): array
    {
        return ['x' => $this->x, 'y' => $this->y];
    }
}

class FlowEdge
{
    public string $id = '';
    public string $source = '';
    public string $target = '';
    public ?string $sourceHandle = null;
    public ?string $targetHandle = null;
    public ?string $label = null;

    public static function fromArray(array $data): self
    {
        $instance = new self();
        $instance->id = $data['id'] ?? '';
        $instance->source = $data['source'] ?? '';
        $instance->target = $data['target'] ?? '';
        $instance->sourceHandle = $data['sourceHandle'] ?? null;
        $instance->targetHandle = $data['targetHandle'] ?? null;
        $instance->label = $data['label'] ?? null;
        return $instance;
    }

    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'source' => $this->source,
            'target' => $this->target,
            'sourceHandle' => $this->sourceHandle,
            'targetHandle' => $this->targetHandle,
            'label' => $this->label,
        ];
    }
}

class FlowSettings
{
    public ?int $timeout = null;
    public int $maxRetries = 3;
    public bool $enableLogging = true;

    public static function fromArray(array $data): self
    {
        $instance = new self();
        $instance->timeout = $data['timeout'] ?? null;
        $instance->maxRetries = $data['maxRetries'] ?? 3;
        $instance->enableLogging = $data['enableLogging'] ?? true;
        return $instance;
    }
}

class FlowExecution
{
    public string $id = '';
    public string $flowId = '';
    public string $status = '';
    public ?string $conversationId = null;
    public ?string $contactId = null;
    public array $variables = [];
    public ?string $currentNodeId = null;
    public array $executionLog = [];
    public ?string $errorMessage = null;
    public ?\DateTimeImmutable $startedAt = null;
    public ?\DateTimeImmutable $completedAt = null;

    public static function fromArray(array $data): self
    {
        $instance = new self();
        $instance->id = $data['id'] ?? '';
        $instance->flowId = $data['flowId'] ?? '';
        $instance->status = $data['status'] ?? '';
        $instance->conversationId = $data['conversationId'] ?? null;
        $instance->contactId = $data['contactId'] ?? null;
        $instance->variables = $data['variables'] ?? [];
        $instance->currentNodeId = $data['currentNodeId'] ?? null;
        $instance->executionLog = $data['executionLog'] ?? [];
        $instance->errorMessage = $data['errorMessage'] ?? null;
        $instance->startedAt = isset($data['startedAt']) ? new \DateTimeImmutable($data['startedAt']) : null;
        $instance->completedAt = isset($data['completedAt']) ? new \DateTimeImmutable($data['completedAt']) : null;
        return $instance;
    }
}

File: sdks/rust/src/types/flow.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum FlowStatus {
    Active,
    Inactive,
    Draft,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum FlowExecutionStatus {
    Running,
    Waiting,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Flow {
    pub id: String,
    pub tenant_id: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub status: FlowStatus,
    pub version: i32,
    #[serde(default)]
    pub nodes: Vec<FlowNode>,
    #[serde(default)]
    pub edges: Vec<FlowEdge>,
    #[serde(default)]
    pub variables: Vec<FlowVariable>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<HashMap<String, serde_json::Value>>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowNode {
    pub id: String,
    #[serde(rename = "type")]
    pub node_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position: Option<HashMap<String, f64>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowEdge {
    pub id: String,
    pub source: String,
    pub target: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_handle: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_handle: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub condition: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowVariable {
    pub name: String,
    #[serde(rename = "type")]
    pub var_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_value: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowExecution {
    pub id: String,
    pub flow_id: String,
    pub conversation_id: String,
    pub status: FlowExecutionStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_node_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<HashMap<String, serde_json::Value>>,
    #[serde(default)]
    pub history: Vec<FlowExecutionStep>,
    pub started_at: chrono::DateTime<chrono::Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<chrono::DateTime<chrono::Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowExecutionStep {
    pub node_id: String,
    pub node_type: String,
    pub started_at: chrono::DateTime<chrono::Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<chrono::DateTime<chrono::Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateFlowInput {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nodes: Option<Vec<FlowNode>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub edges: Option<Vec<FlowEdge>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<Vec<FlowVariable>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

impl CreateFlowInput {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            ..Default::default()
        }
    }

    pub fn description(mut self, desc: impl Into<String>) -> Self {
        self.description = Some(desc.into());
        self
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateFlowInput {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<FlowStatus>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nodes: Option<Vec<FlowNode>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub edges: Option<Vec<FlowEdge>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<Vec<FlowVariable>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExecuteFlowInput {
    pub conversation_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<HashMap<String, serde_json::Value>>,
}

impl ExecuteFlowInput {
    pub fn new(conversation_id: impl Into<String>) -> Self {
        Self {
            conversation_id: conversation_id.into(),
            variables: None,
        }
    }

    pub fn variables(mut self, vars: HashMap<String, serde_json::Value>) -> Self {
        self.variables = Some(vars);
        self
    }
}
